### PR TITLE
improved error message on failed type casting

### PR DESCRIPTION
**Pull request by adrpar**

### CODE REVIEW

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java [hunk 1]
```diff
@@ -590,6 +591,34 @@ public void testInsertFromSubQueryDistributedGroupBy() throws Exception {
     }
 
     @Test
+    public void testInsertFromSubQueryWrongColumnCast() throws Exception {
+        expectedException.expect(SQLActionException.class);
+        expectedException.expectMessage("cannot cast 'foo' to type integer");
+        this.setup.setUpCharacters();
+        execute("create table user (id int, name string)");
+        execute("create table user2 (name string)");
+        ensureYellow();
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: can we test the error cases with a unit test instead?

File: sql/src/main/java/io/crate/operation/scalar/cast/AbstractCastFunction.java [hunk 2]
```diff
@@ -94,8 +95,11 @@ protected Symbol onNormalizeException(Symbol argument, Throwable t) {
     }
 
     protected To onEvaluateException(From argument, Throwable t) {
-        Throwables.propagate(t);
-        return null;
+        String value = argument instanceof BytesRef ? ((BytesRef)argument).utf8ToString() : argument.toString();
+        throw new IllegalArgumentException(
+                String.format(Locale.ENGLISH, "cannot cast '%s' to type %s",
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/scalar/cast/AbstractCastFunction.java]: the error message is now different to the error message that happens in normalizeSymbol (onNormalizeException). And there is also a special ConversionException which is raised from `Literal.convert` which is also used to do casts (if all arguments are literals)

File: core/build.gradle [hunk 3]
```diff
@@ -71,6 +71,7 @@ sourceSets {
             include "io/crate/core/StringUtils*"
             include "io/crate/geo/GeoJSONUtils*"
             include "io/crate/types/**"
+            include "io/crate/exceptions/*"
```
Comment (mfussenegger) [inline core/build.gradle]: why is this necessary? We're only returning SQLActionException to the client and don't want to have the other exception classes on the client, as otherwise we'd break compatibility more often

Comment (adrpar) [inline core/build.gradle]: We need to add a separate exception for validating IP addresses here, so that we can distinguish between cast errors and IP validation errors. Having the IP address validation as ConversionException would just raise a `cannot cast` error otherwise. This might be confusing.

The new exception type has to go into the core, since the types are there and the IP type validates and throws an exception there. Or am I missing something?

File: sql/src/test/java/io/crate/operation/scalar/cast/ToLongFunctionTest.java [hunk 4]
```diff
@@ -96,4 +98,34 @@ public void testEvaluateOnIntAndLong() throws Exception {
...
+        stringFn.evaluate(arg1);
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testEvaluateInvalidByteRef() throws Exception {
+        expectedException.expect(ConversionException.class);
+        expectedException.expectMessage("cannot cast 'hello' to type long");
+        ToPrimitiveFunction stringFn = getFunction(functionName, DataTypes.STRING);
+        Input<Object> arg1 = new Input<Object>() {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/operation/scalar/cast/ToLongFunctionTest.java]: could also use a Literal instead of the Input class

File: sql/src/main/java/io/crate/operation/scalar/cast/AbstractCastFunction.java [hunk 5]
```diff
@@ -80,22 +83,32 @@ public Symbol normalizeSymbol(Function symbol) {
...
+        String value;
+        if (argument instanceof BytesRef) {
+            value = String.format(Locale.ENGLISH, "'%s'", ((BytesRef) argument).utf8ToString());
+        } else if (argument instanceof String) {
+            value = String.format(Locale.ENGLISH, "'%s'", argument);
+        } else {
+            value = argument.toString();
+        }
+        throw new ConversionException(
+                String.format(Locale.ENGLISH, "cannot cast %s to type %s",
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/scalar/cast/AbstractCastFunction.java]: would it make sense to change the ConversionException to take the value and returnType and do the formatting in the ctor to avoid the formatting handling in different places?

Comment (adrpar) [inline sql/src/main/java/io/crate/operation/scalar/cast/AbstractCastFunction.java]: good idea

**Comment by mfussenegger** [comment]
ack

**Comment by mfussenegger** [comment]
ack
